Blocking queue

DemichevD opened the pull request:
Add classes thread-safe class BlockingQueue, Consumers, Producers

DemichevD committed:
Merge pull request #1 from AndreyPlis/master

update

DemichevD committed:
Merge pull request #2 from AndreyPlis/master

update

DemichevD committed:
Add classes thread-safe class BlockingQueue, Consumers, Producers

AndreyPlis commented on src/main/java/com/tibbo/datatable/util/BlockingQueue.java:
```diff
@@ -0,0 +1,104 @@
+    }
+
+    public void remove() {
+        synchronized (this) {
```
Почему здесь синхронизация по this а выше по другому обьекту?

AndreyPlis commented on src/main/java/com/tibbo/datatable/util/BlockingQueue.java:
```diff
@@ -0,0 +1,104 @@
+        }
+    }
+
+    public T element() {
```
Любые обращения к коллекции должно быть синхронизированны!

AndreyPlis commented on src/main/java/com/tibbo/datatable/util/BlockingQueue.java:
```diff
@@ -0,0 +1,104 @@
+        }
+    }
+
+    public void remove() {
```
некорректная сигнатура метода

AndreyPlis commented on src/main/java/com/tibbo/datatable/util/BlockingQueue.java:
```diff
@@ -0,0 +1,104 @@
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+public class BlockingQueue<T> {
```
Лучше заимплементируй стандартный интерфейс BlockingQueue

AndreyPlis commented on src/main/java/com/tibbo/datatable/util/BlockingQueue.java:
```diff
@@ -0,0 +1,104 @@
+package com.tibbo.datatable.util;
```
Нужны тесты. Как понять что эта коллекция работает?

AndreyPlis commented on src/main/java/com/tibbo/datatable/util/BlockingQueue.java:
```diff
@@ -0,0 +1,104 @@
+        synchronized (monitor) {
+            lock.lock();
+            try {
+                while (queue.size() >= capacity) {
```
Циклы плохо. Я говорил на лекции.

AndreyPlis commented on src/main/java/com/tibbo/datatable/util/BlockingQueue.java:
```diff
@@ -0,0 +1,104 @@
+        synchronized (this) {
+            lock.lock();
+            try {
+                while (queue.size() < 1) {
```
Циклы плохо

DemichevD committed:
Add BlockingQueueTest class with tests.
Realized methods: add, offer, put, remove, take, poll, remainingCapacity, element, peek.

DemichevD committed:
Realize test-method in BlockingQueueTest: put, take.

DemichevD committed:
Realize test-method in BlockingQueueTest: queue.

DemichevD committed:
Added some queue tests in BlockingQueueTest.
Fixed thread safety some methods in BlockingQueue

DemichevD committed:
Realize test-method in BlockingQueueTest: threadSafeAdd(), testTakeBlocksWhenEmpty()
